### PR TITLE
feat(config): build_lua + non-default branch rev resolution (closes #97)

### LINKED ISSUE #97: Support Lua function for plugin `build` (needed by blink.cmp v2 and others)
## Motivation

Some modern plugins require a Lua-function build step rather than a plain shell command. The shell-only `build` field is insufficient for them.

Example — blink.cmp v2 emits this warning at startup when the native library is missing:

\`\`\`
blink.cmp  Rust fuzzy matcher not available, falling back to Lua implementation.
blink.cmp  V2 uses a new build/download system for the native library. Please add
            build = function() require('blink.cmp').build():wait(60000) end
            to you lazy.nvim config. See :h blink-cmp-installation for more information.
\`\`\`

The referenced `cmp.build()` does three things (see `blink.cmp/lua/blink/cmp/init.lua`):

1. `cargo build --release`
2. `mv target/release/libblink_cmp_fuzzy.<ext>` → `$data/site/lib/libblink_cmp_fuzzy.<ext>.<short_commit_hash>`
3. `package.loadlib` to verify

Today's rvpm config allows only `build = \"cargo build --release\"`, which covers step 1 only. Steps 2–3 cannot be expressed in a shell one-liner because:

- the destination filename embeds the plugin's current git commit,
- the load-verification must run inside the same Neovim process,
- on Windows MSVC the cargo-produced `blink_cmp_fuzzy.dll` must be renamed to `libblink_cmp_fuzzy.dll` before the move (upstream TODO).

Consequence: users who install blink.cmp via rvpm always fall back to the Lua fuzzy matcher — the Rust native path is unreachable without manually invoking `require('blink.cmp').build()` after every sync / update.

## Proposal

Allow `build` to be either a shell command **or** a Lua callable, matching lazy.nvim's API. Rough sketch:

\`\`\`toml
# shell command (existing behavior)
[[plugins]]
url = \"https://github.com/nvim-treesitter/nvim-treesitter\"
build = \":TSUpdate\"

# Lua function (new) — invoked after the shell command (or instead, if only one is set)
[[plugins]]
url = \"https://github.com/saghen/blink.cmp\"
build_lua = \"require('blink.cmp').build():wait(60000)\"
\`\`\`

A separate field (`build_lua` / `build_cb` / similar) keeps `build` type-stable as `Option<String>` while letting the loader emit a scheduled post-sync Lua call. Alternatively, `build` could become an enum with shell / lua variants.

The Lua callback would need to be:

- run **once per sync/update** of the affected plugin (same trigger as the existing shell `build`),
- run **after** the plugin has been added to rtp (so `require('blink.cmp')` resolves),
- run with the plugin's own working directory (for `debug.getinfo`-based path resolution in blink.cmp).

## Workaround

Users can drop to the Lua-implementation fuzzy matcher (the default fallback) and live with a one-time notification each startup, or run `require('blink.cmp').build():wait(60000)` manually after every `rvpm sync` / `rvpm update`.

## Related

- blink.cmp installation docs: https://cmp.saghen.dev/installation.html
- lazy.nvim `build` spec: https://lazy.folke.io/spec#spec-setup

### PR DESCRIPTION
**Pull request by yukimemi**
## Summary

Closes #97 — adds a parallel `build_lua` field for plugins that need a Lua-function build step (the existing shell-only `build` is insufficient for blink.cmp v2 and similar modern plugins).

```toml
[[plugins]]
url = "https://github.com/saghen/blink.cmp"
build_lua = "require('blink.cmp').build():wait(60000)"
```

## Why

The shell `build` field can only run external commands. blink.cmp v2's installer is a Lua function that:

1. Runs `cargo build --release`
2. Renames the artifact to embed the plugin's current git commit hash
3. Calls `package.loadlib` to verify the install

Steps 2–3 cannot be expressed as a shell one-liner because the destination filename is computed at runtime and the load-verification has to run inside Neovim. The result today: users always fall back to the Lua fuzzy matcher, with a startup warning every time.

## How

- New field `Plugin.build_lua: Option<String>`.
- `execute_build_command` is refactored to run the shell `build` first (preserving existing behavior), then `build_lua` if set.
- `run_build_lua` invokes `nvim --headless -u NONE -l <tmp.lua>`. The script preloads `vim.opt.rtp:append(...)` for the plugin path and its transitive `depends` paths, then runs the user's snippet.
- `cwd` is set to the plugin's clone path; 5-minute timeout matches the shell build.
- `-u NONE` skips user `init.lua` / `vimrc` but keeps the real environment, so `vim.fn.stdpath("data")` etc. resolve to the user's real data dir — this matters for blink.cmp's install destination.
- nvim missing → returns a friendly error pointing at the fix (install nvim, or remove `build_lua`). One failed plugin doesn't abort the whole sync (resilience).

## Test plan

- [x] `cargo make check` — 660 unit tests pass on the new branch (3 new for build_lua / collect_build_rtp).
- [ ] Manual: add blink.cmp with `build_lua = "require('blink.cmp').build():wait(60000)"`, run `rvpm sync`, verify `~/.local/share/nvim/site/lib/libblink_cmp_fuzzy.<ext>.<commit>` exists and the v2 startup warning is gone.
- [ ] Manual: nvim not on PATH → confirm error message points at the fix and other plugins still complete.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added a new build_lua option to run Lua snippets as a post-build step; build now supports two-stage execution (shell then Lua).

* **Documentation**
  * Expanded plugin docs: build runs after clone/update, uses headless Neovim, includes transitive runtime paths, 5‑minute timeout, and non-fatal failure reporting; details on data-path resolution and accepted snippet formats.

* **Bug Fixes / Reliability**
  * More reliable branch/revision resolution during clone/fetch so non-default revisions sync correctly.

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -614,6 +647,87 @@ async fn execute_build_command(
...
+
+    let build_timeout = std::time::Duration::from_secs(300);
+    let child = match tokio::process::Command::new("nvim")
+        .args(["--headless", "-u", "NONE", "-l"])
+        .arg(&tmp_path)
+        .current_dir(dst_path)
+        .stdout(std::process::Stdio::piped())
+        .stderr(std::process::Stdio::piped())
+        .spawn()
+    {
```
[the remark this answers is not in the document]
Comment (yukimemi) [reply src/main.rs:757]: @gemini-code-assist Fixed in ee9c239 — `.kill_on_drop(true)` is now set on the nvim Command so timeout-fired future drops kill the child.

File: src/main.rs [hunk 2]
```diff
@@ -559,15 +559,38 @@ fn resolve_plugin_dst(plugin: &crate::config::Plugin, cache_root: &Path) -> Path
...
+/// 設定されていれば、shell → lua の順に実行する。どちらか一方でも失敗したら
+/// 即座にエラー文字列を返す。両方未設定なら `None`。
 async fn execute_build_command(
     plugin: &crate::config::Plugin,
     dst_path: &Path,
     config: &crate::config::Config,
     cache_root: &Path,
 ) -> Option<String> {
-    let build_cmd = plugin.build.as_ref()?;
+    let rtp_dirs = collect_build_rtp(plugin, dst_path, config, cache_root);
```
[the remark this answers is not in the document]
Comment (yukimemi) [reply src/main.rs:577]: @gemini-code-assist Fixed in ee9c239 — `execute_build_command` now early-returns at the top if both `build` and `build_lua` are `None`, skipping the transitive depends scan entirely.

File: src/main.rs [hunk 3]
```diff
@@ -614,6 +647,87 @@ async fn execute_build_command(
...
+        "rvpm-build-{}-{}.lua",
+        std::process::id(),
+        std::time::SystemTime::now()
+            .duration_since(std::time::UNIX_EPOCH)
+            .map(|d| d.as_nanos())
+            .unwrap_or(0)
+    ));
+    if let Err(e) = std::fs::write(&tmp_path, &script) {
+        return Some(format!("build_lua: failed to write temp script: {e}"));
+    }
```
[the remark this answers is not in the document]
Comment (yukimemi) [reply src/main.rs]: @gemini-code-assist Fixed in ee9c239 — switched to `tempfile::Builder::new().prefix("rvpm-build-").suffix(".lua").tempfile()`. `NamedTempFile` auto-deletes on drop so signal interruption no longer leaves orphans in $TMPDIR.

**Comment by yukimemi** [comment]
Two more pushes since the previous review:

**ee9c239** — applied 3 Gemini findings (kill_on_drop, early-return, tempfile crate). All correctness/leak fixes per the durable bug-fix policy.

**0fcb576** — owner spotted while testing: the `should_build` gate in `run_sync` only checked `plugin.build.is_some()`, so a plugin with only `build_lua` set would silently skip its build step on `rvpm sync` / `--rebuild` / TUI `R`. Fixed gate + made the status label show `Building: <cmd> + (lua)` / `Building: (lua build)` when relevant.

@gemini-code-assist @coderabbitai please take another look.

**Comment by yukimemi** [comment]
Owner reported on the branch: `rvpm sync --rebuild blink.cmp` was failing with E5112 because they wrote `build_lua = "function() require(...).build():wait(60000) end"` (lazy.nvim style), and `function()` in Lua statement context expects a name.

**8e6055f** auto-unwraps the lazy.nvim wrapper at the top of `run_build_lua`:
- `function() X end` → `X` (unwrapped)
- `X` (plain statement) → unchanged
- `function name() ... end` (named decl) → unchanged (already valid)

6 new unit tests for the unwrap (lazy.nvim style, space variant, plain statement no-op, named decl no-op, identifier ending in `end`, nested function blocks).

README updated to mention both accepted forms. 666 tests passing.

@gemini-code-assist @coderabbitai please re-review.

File: src/main.rs [hunk 4]
```diff
@@ -614,6 +653,125 @@ async fn execute_build_command(
...
+    let tmp_file = match tempfile::Builder::new()
+        .prefix("rvpm-build-")
+        .suffix(".lua")
+        .tempfile()
+    {
+        Ok(f) => f,
+        Err(e) => return Some(format!("build_lua: failed to create temp script: {e}")),
+    };
+    let tmp_path = tmp_file.path().to_path_buf();
+    if let Err(e) = std::fs::write(&tmp_path, &script) {
```
[the remark this answers is not in the document]
Comment (yukimemi) [reply src/main.rs]: @gemini-code-assist Fixed in 9a6776a — switched the temp-script write to `tokio::fs::write(...).await` so the executor isn't blocked while we wait on disk I/O. Concurrent plugin builds in `run_sync` benefit accordingly.

**Comment by yukimemi** [comment]
@coderabbitai a couple more commits since your last analysis (8e6055f):

**1ef481b** — refactored the unwrap helper to a regex (`(?s)^\s*function\s*\(\s*\)\s*(.*?)\s*end\s*(?:--[^\n\r]*)?\s*$`). This adds proper multi-line support, trailing line-comment tolerance, and whitespace flexibility around `function ( )`. The previous byte-level strip already handled multi-line correctly but had edge cases. 6 new tests cover multi-line bodies, nested `function() ... end` (preserved verbatim — only the outer wrapper is unwrapped), trailing comment after `end`, and a negative test for `function(args)` (intentionally not unwrapped — only zero-arity wrappers).

**9a6776a** — applied your sibling Gemini Medium suggestion: switched the temp-script write from `std::fs::write` (blocking the tokio executor) to `tokio::fs::write(...).await`. Concurrent plugin builds in `run_sync` no longer stall sibling tasks during disk I/O.

671 unit tests passing, CI green. Would you take a look?

**Comment by yukimemi** [comment]
Two git-side fixes added to this PR while testing build_lua against blink.cmp v1:

**d958963** — gix's prepare_clone writes a narrow refspec (`refs/heads/<default>:refs/remotes/origin/<default>`), so non-default branches like `v1` are never fetched into remote tracking refs. Even with `rev = "v1"` set in `config.toml`, the next `rvpm sync` would error `rev not found`.

Fix: a small `ensure_all_branches_refspec` helper rewrites the `[remote "origin"].fetch` entry in `.git/config` to git CLI's default `+refs/heads/*:refs/remotes/origin/*`. Called once after clone and again before every fetch, idempotent, and only touches the `[remote "origin"]` section so other remotes are preserved. Existing clones get healed automatically on the next sync.

**47dab8f** — even with the wider refspec landing all branches as `refs/remotes/origin/*`, `gix::rev_parse_single("v1")` follows git's short-name resolution (refs/heads/, refs/tags/, then SHA) and does NOT auto-search `refs/remotes/origin/`. git CLI's `git checkout v1` would create an auto-tracking local branch in this case; gix needs an explicit lookup.

Fix: when `rev_parse_single` fails, `gix_checkout` now falls back to looking up `refs/remotes/origin/<rev>` and creates the local branch pointing at the resolved commit. Mirrors `git checkout` first-time auto-tracking, idempotent on subsequent syncs (just updates the local branch tip).

End-to-end test (`test_sync_resolves_non_default_branch_via_full_refspec`) confirms the full flow: a non-default branch `v1` set as `rev` is fetched and checked out cleanly. Three additional unit tests cover the refspec normalization (replace, idempotent, only-origin-section).

675 unit tests passing, CI green. @gemini-code-assist @coderabbitai please take a look — these are git-flow fixes alongside the original build_lua scope, but I figured it's cleaner than splitting into a separate PR since the user reproducer that surfaced both was the same blink.cmp v1 install path.

File: src/git.rs [hunk 5]
```diff
@@ -445,21 +456,113 @@ fn fetch_impl(dst: &Path) -> Result<()> {
...
+    // [remote "origin"] セクションそのものが無い (or fetch 行が無い) ケースでは
+    // 末尾に追記する。普通の clone なら必ずあるはずだがガード。
+    if !replaced {
+        if !new_content.contains("[remote \"origin\"]") {
+            new_content.push_str("[remote \"origin\"]\n");
+        }
+        new_content.push_str("\tfetch = ");
+        new_content.push_str(want);
+        new_content.push('\n');
+    }
```
[the remark this answers is not in the document]
Comment (yukimemi) [reply src/git.rs]: @gemini-code-assist Fixed in 77cdd43 — `ensure_all_branches_refspec` now tracks origin scope during iteration: when the NEXT section header (or EOF) arrives without having seen a `fetch` line inside origin, the line is injected at that boundary so it stays in origin's scope. New test `ensure_all_branches_refspec_inserts_into_origin_when_origin_is_not_last_section` covers a config where `[branch "main"]` follows `[remote "origin"]`.

File: src/git.rs [hunk 6]
```diff
@@ -427,10 +427,21 @@ fn clone_impl(url: &str, dst: &Path) -> Result<()> {
             anyhow::anyhow!("checkout failed: {}", e)
         })?;
 
+    // clone 直後に refspec を全 branch に正規化しておくと、user が `rev = "v1"`
+    // 等の非デフォルト branch を指定したケースで次回 fetch から拾える。
+    let _ = ensure_all_branches_refspec(dst);
```
[the remark this answers is not in the document]
Comment (yukimemi) [reply src/git.rs]: @gemini-code-assist Fixed in 77cdd43 — switched `let _ = ...` to `?` so clone-time refspec write failures propagate. Silent IO error would have surfaced later as confusing "rev not found" on non-default branches.

File: src/git.rs [hunk 7]
```diff
@@ -445,21 +456,113 @@ fn fetch_impl(dst: &Path) -> Result<()> {
...
+    // 拾ったケースのどちらでも、symbolic HEAD で local branch を立てる。
+    // (2) のとき local branch がまだ無ければ作る (= git CLI の `checkout <branch>`
+    // で自動 tracking branch を作るのと同じ振る舞い)。
     let branch_ref = format!("refs/heads/{}", rev);
-    if repo.find_reference(&branch_ref).is_ok() {
-        // HEAD を symbolic ref にする (直接ファイル書き込み)
+    let local_branch_exists = repo.find_reference(&branch_ref).is_ok();
+    let should_set_branch = local_branch_exists
+        || matches!(source, DirectOrRemote::FromRemote)
+        || matches!(source, DirectOrRemote::Direct) && repo.find_reference(&branch_ref).is_ok();
```
[the remark this answers is not in the document]
Comment (yukimemi) [reply src/git.rs]: @gemini-code-assist Fixed in 77cdd43 — `should_set_branch` is now `local_branch_exists || matches!(source, DirectOrRemote::FromRemote)`. The redundant `Direct && exists` clause was equivalent to `local_branch_exists` already.

File: src/git.rs [hunk 8]
```diff
@@ -1036,6 +1189,186 @@ mod tests {
...
+        let url = format!("file://{}", src.display());
+        let repo = Repo::new(&url, &dst, Some("v1"));
+        repo.sync()
+            .await
+            .expect("sync to v1 should succeed after refspec normalization");
+
+        // v1 の HEAD に揃っていること
+        let head = repo.head_commit().await.unwrap();
+        assert_eq!(head, v1_head, "checkout should land on v1 tip");
+    }
```
[the remark this answers is not in the document]
Comment (yukimemi) [reply src/git.rs:1374]: @coderabbitai Excellent catch — the test was passing for the wrong reason. After applying your suggested setup (capture the init default branch right after `git init`, restore it before clone), the test correctly failed with the same "rev not found" error the user originally reported. That surfaced an additional bug: in the **fresh-clone** path of `sync_impl`, `clone_impl` writes a narrow refspec (only the default branch); my earlier `ensure_all_branches_refspec` widens `.git/config` but only takes effect on the NEXT fetch. The fresh-clone path then went straight to `gix_checkout(dst, rev)` without re-fetching, so `refs/remotes/origin/v1` was never populated.

Fixed in d183ab4 — sync_impl now calls `fetch_impl(dst)` before `gix_checkout(dst, rev)` on fresh clones when a rev is specified. fetch_impl is idempotent and calls ensure_all_branches_refspec at the top, so the second fetch pulls all branches with the wide refspec. No-rev clones are unchanged.

The regression test now genuinely exercises the `refs/remotes/origin/<rev>` path AND the fresh-clone re-fetch step. 678 unit tests passing.
